server/admin/tree_gc_test.go

```diff
@@ -175,6 +177,176 @@ func TestDeletedTreeGC_RunOnce(t *testing.T) {
+		},
+		{
+			desc:        "commitErr",
+			trees:       []*trillian.Tree{logTree1, logTree2},
+			beginErrs:   []error{nil, nil},
+			deleteErrs:  []error{nil, nil},
+			commitErrs:  []error{errors.New("commit err"), nil},
+			wantTreeIDs: []int64{logTree1.TreeId, logTree2.TreeId},
+			wantCount:   1,
+			wantErrs:    []string{"commit err"},
+		},
+		{
+			// logTree1 = delete successful
+			// logTree2 = delete error
+			// mapTree  = commit error
+			// badTS    = timestamp parse error (no HardDeleteTree() call)
+			desc:        "multipleErrors",
+			trees:       []*trillian.Tree{logTree1, logTree2, mapTree, badTS},
+			beginErrs:   []error{nil, nil, nil},
+			deleteErrs:  []error{nil, errors.New("delete err"), nil},
+			commitErrs:  []error{nil, nil, errors.New("commit err")},
+			wantTreeIDs: []int64{logTree1.TreeId, logTree2.TreeId, mapTree.TreeId},
+			wantCount:   1,
+			wantErrs:    []string{"delete err", "commit err", "error parsing delete_time"},
+		},
+	}
+
+	ctx := context.Background()
+	for _, test := range tests {
+		as := storage.NewMockAdminStorage(ctrl)
+
+		listTX := storage.NewMockReadOnlyAdminTX(ctrl)
+		lastTXCall := as.EXPECT().Snapshot(gomock.Any()).Return(listTX, test.snapshotErr)
+		listTX.EXPECT().ListTrees(gomock.Any(), true /* includeDeleted */).AnyTimes().Return(test.trees, test.listErr)
+		listTX.EXPECT().Close().AnyTimes().Return(nil)
+		listTX.EXPECT().Commit().AnyTimes().Return(test.snapshotCommitErr)
+
+		// Sanity check test setup
+		if l1, l2, l3, l4 := len(test.beginErrs), len(test.deleteErrs), len(test.commitErrs), len(test.wantTreeIDs); l1 != l2 || l1 != l3 || l1 != l4 {
+			t.Fatalf("%v: beginErrs, deleteErrs, commitErrs and wantTreeIDs have different lenghts: %v, %v, %v and %v", test.desc, l1, l2, l3, l4)
+		}
+
+		for i, beginErr := range test.beginErrs {
+			deleteErr := test.deleteErrs[i]
+			commitErr := test.commitErrs[i]
+			id := test.wantTreeIDs[i]
+
+			deleteTX := storage.NewMockAdminTX(ctrl)
+			lastTXCall = as.EXPECT().Begin(gomock.Any()).Return(deleteTX, beginErr).After(lastTXCall)
+			deleteTX.EXPECT().HardDeleteTree(gomock.Any(), id).AnyTimes().Return(deleteErr)
+			deleteTX.EXPECT().Close().AnyTimes().Return(nil)
+			deleteTX.EXPECT().Commit().AnyTimes().Return(commitErr)
+		}
+
+		gc := NewDeletedTreeGC(
+			as, deleteThreshold, 1*time.Second /* minRunInterval */, nil /* mf */)
+		count, err := gc.RunOnce(ctx)
+		if err == nil {
+			t.Errorf("%v: RunOnce() returned err = nil, want non-nil", test.desc)
+			continue
+		}
+
+		if count != test.wantCount {
+			t.Errorf("%v: RunOnce() = %v, want = %v", test.desc, count, test.wantCount)
+		}
+
+		failed := false
+		for _, want := range test.wantErrs {
+			if !strings.Contains(err.Error(), want) {
+				if !failed {
+					t.Errorf("%v: RunOnce() returned err = %v (see following errors)", test.desc, err)
+					failed = true
+				}
+				t.Errorf("%v: RunOnce(): err doesn't contain %q", test.desc, want)
```

Comment:
You could just include `err` in this log and drop the `failed` stuff.  Or not, up to you.

Reply:
Done. I thought it would look terrible, but it isn't that bad tbh.